### PR TITLE
[serve] Re-add variable route support for old API

### PR DESCRIPTION
**Pull request by edoakes**
## Why are these changes needed?

This was mistakenly removed in https://github.com/ray-project/ray/pull/15193 but we want to maintain full backwards compatibility at least through Ray 1.4.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/http_proxy.py [hunk 1]
```diff
@@ -147,6 +273,22 @@ async def _not_found(self, scope, receive, send):
...
+        if route_prefix is None:
+            return await self._not_found(scope, receive, send)
+
+        # Modify the path and root path so that reverse lookups and redirection
+        # work as expected. We do this here instead of in replicas so it can be
+        # changed without restarting the replicas.
+        if route_prefix != "/":
+            assert not route_prefix.endswith("/")
+            scope["path"] = scope["path"].replace(route_prefix, "", 1)
+            scope["root_path"] = route_prefix
```
Comment (architkulkarni) [inline python/ray/serve/http_proxy.py:288]: Should this be covered by a test that does reverse lookups and redirection (or is there already such a test?)

**Review by architkulkarni — APPROVED** [review]
Looks great!

File: python/ray/serve/http_proxy.py (hunk 1 above)
Comment (simon-mo) [reply python/ray/serve/http_proxy.py:288]: Ah I think I wrote this comment. You can add a test for `url_path_for` https://stackoverflow.com/questions/63682956/fastapi-retrieve-url-from-view-name-route-name for something...

Comment (edoakes) [reply python/ray/serve/http_proxy.py:288]: Adding a test for this in a separate PR because it's unrelated https://github.com/ray-project/ray/pull/15461

**Review by simon-mo — APPROVED** [review]

**Pull request merged**